src/cascadia/TerminalApp/VsSetupConfiguration.cpp

```diff
@@ -0,0 +1,101 @@
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "VsSetupConfiguration.h"
+
+std::vector<TerminalApp::VsSetupConfiguration::VsSetupInstance> TerminalApp::VsSetupConfiguration::QueryInstances()
+{
+    std::vector<VsSetupConfiguration::VsSetupInstance> instances;
+
+    // SetupConfiguration is only registered if Visual Studio is installed
+    ComPtrSetupQuery pQuery{ wil::CoCreateInstanceNoThrow<SetupConfiguration, ISetupConfiguration2>() };
+    if (pQuery == nullptr)
+    {
+        return instances;
+    }
+
+    // Enumerate all valid instances of Visual Studio
+    wil::com_ptr<IEnumSetupInstances> e;
+    THROW_IF_FAILED(pQuery->EnumInstances(&e));
+
+    std::array<ComPtrSetupInstance, 1> rgpInstance;
+    auto result = e->Next(1, &rgpInstance[0], NULL);
+    while (S_OK == result)
+    {
+        // wrap the COM pointers in a friendly interface
+        instances.emplace_back(VsSetupConfiguration::VsSetupInstance{ pQuery, rgpInstance[0] });
+        result = e->Next(1, &rgpInstance[0], NULL);
+    }
+
+    THROW_IF_FAILED(result);
+
+    return instances;
+}
+
+/// <summary>
+/// Takes a relative path under a Visual Studio installation and returns the absolute path.
+/// </summary>
+std::wstring TerminalApp::VsSetupConfiguration::ResolvePath(ComPtrSetupInstance pInst, std::wstring_view relativePath)
+{
+    wil::unique_bstr bstrAbsolutePath;
+    THROW_IF_FAILED(pInst->ResolvePath(relativePath.data(), &bstrAbsolutePath));
+    return bstrAbsolutePath.get();
+}
+
+/// <summary>
+/// Determines whether a Visual Studio installation version falls within a specified range.
+/// The range is specified as a string, ex: "[15.0.0.0,)", "[15.0.0.0, 16.7.0.0)
+/// </summary>
+bool TerminalApp::VsSetupConfiguration::InstallationVersionInRange(ComPtrSetupQuery pQuery, ComPtrSetupInstance pInst, std::wstring_view range)
+{
+    auto helper = pQuery.query<ISetupHelper>();
+
+    // VS versions in a string format such as "16.3.0.0" can be easily compared
+    // by parsing them into 64-bit unsigned integers using the stable algorithm
+    // provided by ParseVersion and ParseVersionRange
+
+    unsigned long long minVersion{ 0 };
+    unsigned long long maxVersion{ 0 };
+    THROW_IF_FAILED(helper->ParseVersionRange(range.data(), &minVersion, &maxVersion));
+
+    wil::unique_bstr bstrVersion;
+    THROW_IF_FAILED(pInst->GetInstallationVersion(&bstrVersion));
+
+    unsigned long long ullVersion{ 0 };
+    THROW_IF_FAILED(helper->ParseVersion(bstrVersion.get(), &ullVersion));
+
+    return ullVersion >= minVersion && ullVersion <= maxVersion;
+}
+
+std::wstring TerminalApp::VsSetupConfiguration::GetInstallationVersion(ComPtrSetupInstance pInst)
+{
+    wil::unique_bstr bstrInstallationVersion;
+    THROW_IF_FAILED(pInst->GetInstallationVersion(&bstrInstallationVersion));
+    return bstrInstallationVersion.get();
+}
+
+std::wstring TerminalApp::VsSetupConfiguration::GetInstallationPath(ComPtrSetupInstance pInst)

```

Comment:
What is this used for? Given your use of parsing versions and resolving paths, I wouldn't think this necessary.

Reply:
It's used to set the default start-in path for Developer PowerShell. Developer PowerShell has some interesting logic for setting the start-in path. By putting the path in the terminal profile, and disabling the internal logic, the user's terminal profile determines the start-in path.

Reply:
Why does starting up in the installation root of Visual Studio make sense? Starting in a user directory from which project files are easily navigable makes much more sense, and `%userprofile%` - the seeming default - is closer to that than something deep under Program Files.

Reply:
It's where Developer Command Prompt starts